### PR TITLE
Support grndb --log-flags

### PR DESCRIPTION
**Pull request by kenhys**

### CODE REVIEW

**Comment by kenhys** [comment]
If https://github.com/kou/mruby-slop/pull/1 is merged, this PR works as expected.

File: test/command_line/suite/grndb/test_options.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+class TestGrnDBOptions < GroongaTestCase
+  sub_test_case("--log-flags") do
+
```
[the remark this answers is not in the document]
Comment (kenhys) [reply test/command_line/suite/grndb/test_options.rb]: Thanks. I've added it too.

**Comment by kenhys** [comment]
@komainu8 I've rebased and added default test case. could you review again please?

**Comment by kenhys** [comment]
now, mruby-slop supports slop 4.6.2. https://github.com/kou/mruby-slop/commit/0aa5b832315ccd1ddc55f5391f2d8a1e9d6145b0

**Comment by kenhys** [comment]
To use latest mruby-slop, need to update vendor/mruby-source, but I've found that there is some incompatible changes are introduced.

```
/usr/bin/ld: ./.libs/libgroonga.so: undefined reference to `MKOP_A'
/usr/bin/ld: ./.libs/libgroonga.so: undefined reference to `GET_OPCODE'
/usr/bin/ld: ./.libs/libgroonga.so: undefined reference to `MKOP_AB'
collect2: error: ld returned 1 exit status
```

This incompatibility is introduced at https://github.com/mruby/mruby/commit/891839b976c75c77f238931123ac472e3284e95d

The feature about keyword argument is supported after above commit, so need to support new bytecode implementation.

https://github.com/mruby/mruby/commits/stable?before=8ee516436b8d174a50764939bee23a442aa00b3f+661

**Comment by kenhys** [comment]
And recent mruby-time split header file, so include path need to be updated build mruby.

```
diff --git a/vendor/mruby/Makefile.am b/vendor/mruby/Makefile.am
index c7c021ac4..e330469a2 100644
--- a/vendor/mruby/Makefile.am
+++ b/vendor/mruby/Makefile.am
@@ -11,7 +11,8 @@ DEFAULT_INCLUDES =                                            \
        -I$(srcdir)/../mruby-source/include                     \
        -I$(srcdir)/../mruby-source/src                         \
        -I$(srcdir)/../mruby-source/mrbgems/mruby-compiler/core \
-       -I$(srcdir)/../mruby-source/mrbgems/mruby-io/include
+       -I$(srcdir)/../mruby-source/mrbgems/mruby-io/include    \
+       -I$(srcdir)/../mruby-source/mrbgems/mruby-time/include
 
```

**Comment by kenhys** [comment]
even though on only travis fails, now test passes on locally.

**Comment by kenhys** [comment]
Could you review this PR, @kou

**Comment by kou** [comment]
Could you rebase on master?
Most of CI failures will be fixed.

File: test/command_line/suite/grndb/test_options.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (kou) [inline test/command_line/suite/grndb/test_options.rb]: Could you remove this empty line?

File: lib/mrb/scripts/logger/flags.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+      attr_reader :name, :flag
+      def initialize(name, flag)
+        @@flags[flag] = self
+        @name  = name
+        @flag = flag
+      end
+
+      def to_s
+        @@flags_names
```
Comment (kou) [inline lib/mrb/scripts/logger/flags.rb]: Why `@name` isn't used?

File: lib/mrb/scripts/logger/flags.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    class Flags
+      @@flags = {}
+      @@flag_names = nil
+      class << self
+        def set_flag_names(flag)
+          flag_names = @@flags.collect do |index, candidate|
+            candidate.name if candidate.flag & flag > 0
+          end
+          @@flag_names = flag_names.compact.join("|")
+        end
```
Comment (kou) [inline lib/mrb/scripts/logger/flags.rb]: `format` or something is better.

File: lib/mrb/scripts/logger/flags.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+module Groonga
+  class Logger
+    class Flags
+      @@flags = {}
+      @@flag_names = nil
```
Comment (kou) [inline lib/mrb/scripts/logger/flags.rb]: Is this needed?

File: lib/mrb/scripts/logger/flags.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+module Groonga
+  class Logger
+    class Flags
```
Comment (kou) [inline lib/mrb/scripts/logger/flags.rb]: Why `Flags` not `Flag`?

File: test/command_line/suite/grndb/test_options.rb (hunk 2 above)
Comment (kenhys) [reply test/command_line/suite/grndb/test_options.rb]: I've fixed.

File: lib/mrb/scripts/logger/flags.rb (hunk 4 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: I've fixed.

File: lib/mrb/scripts/logger/flags.rb (hunk 5 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: No, It can be removed.

File: lib/mrb/scripts/logger/flags.rb (hunk 6 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: Oh, for consistency, it seems that `Flag` is better.

Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: I've fixed.

File: lib/mrb/scripts/logger/flags.rb (hunk 3 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: `@name` is implicitly used when correcting each flag name.

Comment (kenhys) [reply lib/mrb/scripts/logger/flags.rb]: (`@name` is not enough to collect all default flags, so I've used `@flags_names` in the previous version)

**Comment by kenhys** [comment]
@kou Thanks, I've fixed based on your feedback.

File: src/grndb.c [hunk 7]
```diff
@@ -183,6 +195,17 @@ main(int argc, char **argv)
     grn_default_logger_set_max_level(log_level);
   }
 
+  int log_flags = GRN_LOG_DEFAULT;
+  grn_default_logger_set_flags(log_flags);
```
Comment (kou) [inline src/grndb.c]: Why is this needed?

File: test/command_line/suite/grndb/test_options.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      grndb("check")
+      assert_equal("1970-01-01 00:00:00.000000|n| " +
+                   "grn_init: <VERSION>",
+                   normalize_init_line(File.readlines(@log_path).first))
+    end
+
+    test("add: one") do
+      flags = "+pid"
+      groonga("status",
+              command_line: ["--log-flags", flags])
```
Comment (kou) [inline test/command_line/suite/grndb/test_options.rb]: Why is this option is needed for `groonga`?

File: lib/mrb/scripts/command_line_parser.rb [hunk 9]
```diff
@@ -97,6 +97,9 @@ def setup_common_options(options)
       options.string("--log-level",
                      "Change log level (#{default_log_level.name})",
                      default: default_log_level)
+      default_log_flags = Logger.default_flags
+      options.string("--log-flags",
+                     "Change log flags (#{default_log_flags.format})")
```
Comment (kou) [inline lib/mrb/scripts/command_line_parser.rb]: Why is`default:` not used?

File: lib/mrb/mrb_logger.c [hunk 10]
```diff
@@ -57,6 +57,18 @@ logger_s_get_default_level(mrb_state *mrb, mrb_value self)
...
 
+static mrb_value
+logger_s_get_default_flags(mrb_state *mrb, mrb_value self)
+{
+  mrb_value mrb_flags_class;
+  mrb_value mrb_flags;
+
+  mrb_flags_class = mrb_const_get(mrb, self, mrb_intern_lit(mrb, "Flag"));
+  mrb_flags = mrb_fixnum_value(grn_default_logger_get_flags());
+  return mrb_funcall(mrb, mrb_flags_class, "set_flags", 1, mrb_flags);
```
Comment (kou) [inline lib/mrb/mrb_logger.c]: Calling setter in `default_XXX` isn't good API.

File: lib/mrb/mrb_logger.c [hunk 11]
```diff
@@ -57,6 +57,18 @@ logger_s_get_default_level(mrb_state *mrb, mrb_value self)
...
+}
+
+
```
Comment (kou) [inline lib/mrb/mrb_logger.c]: Could you remove an empty needless line?

File: lib/mrb/scripts/logger/flag.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+      NONE       = new(:none,       0)
+      TIME       = new(:time,       1)
+      TITLE      = new(:title,      2)
+      MESSAGE    = new(:message,    4)
+      LOCATION   = new(:location,   8)
+      PID        = new(:pid,        16)
+      PROCESS_ID = new(:process_id, 16)
+      THREAD_ID  = new(:thread_id,  32)
+    end
```
Comment (kou) [inline lib/mrb/scripts/logger/flag.rb]: Umm. This class isn't a good API.

How about the following?

```ruby
module Groonga
  class Logger
    class Flags
      VALUES = {
        time: 1,
        title: 2,
        message: 4,
        location: 8,
        process_id: 16,
        thread_id: 32,
      }

      attr_reader :value
      def initialize(value)
        case value
        when String, Symbol
          @value = VALUES[value.to_sym]
          if @value.nil?
            available_names = VALUES.keys.inspect
            message = "unknown flag name: #{value.inspect}: #{available_names}"
            raise ArgumentError, message
          end
        else
          @value = value
        end
      end

      def to_i
        @value
      end

      def to_s
        names = []
        VALUES.each do |name, value|
          names << name.to_s unless (value & @value) == 0
        end
        names.join("|")
      end

      def |(other)
        other = self.class.new(other) unless other.is_a?(self.class)
        new(@value | other.to_i)
      end

      def &(other)
        other = self.class.new(other) unless other.is_a?(self.class)
        new(@value & other.to_i)
      end

      NONE = new(0)
      TIME = new(:time)
      TITLE = new(:title)
      MESSAGE = new(:message)
      LOCATION = new(:location)
      PROCESS_ID = new(:process_id)
      PID = PROCESS_ID
      THREAD_ID = new(:thread_id)
    end
  end
end
```

```c
static mrb_value
logger_s_get_default_flags(mrb_state *mrb, mrb_value self)
{
  mrb_value mrb_flags_class =
    mrb_const_get(mrb, self, mrb_intern_lit(mrb, "Flag"));
  mrb_value mrb_flags = mrb_fixnum_value(grn_default_logger_get_flags());
  return mrb_funcall(mrb, mrb_flags_class, "new", 1, mrb_flags);
}
```

File: src/grndb.c (hunk 7 above)
Comment (kenhys) [reply src/grndb.c]: Oh, It is needless. fixed.

File: test/command_line/suite/grndb/test_options.rb (hunk 8 above)
Comment (kenhys) [reply test/command_line/suite/grndb/test_options.rb]: Ah, It is needless. And I've noticed that test case is wrong. I've fixed.

File: lib/mrb/mrb_logger.c (hunk 11 above)
Comment (kenhys) [reply lib/mrb/mrb_logger.c]: fixed.

File: lib/mrb/scripts/logger/flag.rb (hunk 12 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flag.rb]: It seems that it is reasonable API which I couldn't design.

**Comment by kenhys** [comment]
@kou I've fixed based on your suggestion. could you review again?

File: lib/mrb/scripts/logger/flag.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
+module Groonga
+  class Logger
+    class Flag
```
Comment (kou) [inline lib/mrb/scripts/logger/flag.rb]: In this API, this should be `Flags`. Because this object has multiple flag values. For example, `Flags.new(1 | 2 | 4)` is valid for `time`, `title` and `message` flags.

File: src/grndb.c [hunk 14]
```diff
@@ -184,6 +196,16 @@ main(int argc, char **argv)
     grn_default_logger_set_max_level(log_level);
   }
 
+  int log_flags = GRN_LOG_DEFAULT;
```
Comment (kou) [inline src/grndb.c]: Could you reduce scope by putting this into `if (log_flags_name)` block?

Comment (kou) [reply src/grndb.c]: We don't need to initialize this by `GRN_LOG_DEFAULT`. Because the default value isn't used.

File: lib/mrb/scripts/logger/flag.rb (hunk 13 above)
Comment (kenhys) [reply lib/mrb/scripts/logger/flag.rb]: I've got it.

File: src/grndb.c (hunk 14 above)
Comment (kenhys) [reply src/grndb.c]: Fixed.

**Comment by kenhys** [comment]
@kou I've fixed them!
